### PR TITLE
Making upgrading mantis clusters graceful and mindful of job's SLA constraints

### PR DESCRIPTION
**Pull request by sundargates**
### Context

Explain context and other details for this pull request.

### Checklist

- [ ] `./gradlew build` compiles code correctly
- [ ] Added new tests where applicable
- [ ] `./gradlew test` passes all tests
- [ ] Extended README or added javadocs where applicable
- [ ] Added copyright headers for new files from `CONTRIBUTING.md`

### CODE REVIEW

**Comment by Andyz26** [comment]
Can you add a brief description to PR?

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java [hunk 1]
```diff
@@ -77,6 +79,12 @@ public interface ResourceCluster extends ResourceClusterGateway {
 
     CompletableFuture<TaskExecutorStatus> getTaskExecutorState(TaskExecutorID taskExecutorID);
 
+    CompletableFuture<Void> disabledTaskExecutorsAre(Map<String, String> attributes, Instant expiry);
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java:82]: Is this an action or query? Can you update the name or add doc?

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/TaskExecutorRegistration.java [hunk 2]
```diff
@@ -40,4 +41,11 @@ public class TaskExecutorRegistration {
 
   // machine information identifies the cpu/mem/disk/network resources of the task executor.
   MachineDefinition machineDefinition;
+
+  // custom attributes describing the task executor
+  Map<String, String> taskExecutorAttributes;
+
+  public boolean containsAttributes(Map<String, String> requiredAttributes) {
+      return taskExecutorAttributes.entrySet().containsAll(requiredAttributes.entrySet());
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/TaskExecutorRegistration.java:49]: Null check? in #239 I didn't mark this one as nonNull yet for back-compat on existing workers.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/api/akka/route/v1/ResourceClustersNonLeaderRedirectRoute.java [hunk 3]
```diff
@@ -233,6 +242,13 @@ private Route getTaskExecutorState(ClusterID clusterID, TaskExecutorID taskExecu
         return withFuture(statusOverview);
     }
 
+    private Route disableTaskExecutors(ClusterID clusterID) {
+        return entity(Jackson.unmarshaller(DisableTaskExecutorsRequest.class), request -> {
+            log.info("POST /api/v1/resourceClusters/{}/disableTaskExecutors called with body {}", clusterID, request);
+            return withFuture(gateway.getClusterFor(clusterID).disabledTaskExecutorsAre(request.getAttributes(), Instant.now().plus(Duration.ofHours(24))));
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/api/akka/route/v1/ResourceClustersNonLeaderRedirectRoute.java:248]: Can you add some doc about how/why this disable API with duration is used?

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterAkkaImpl.java [hunk 4]
```diff
@@ -170,4 +172,21 @@ public CompletableFuture<TaskExecutorStatus> getTaskExecutorState(TaskExecutorID
...
+    @Override
+    public CompletableFuture<Void> disabledTaskExecutorsAre(Map<String, String> attributes, Instant expiry) {
+        resourceClusterManagerActor.tell(
+            new ResourceClusterActor.DisableTaskExecutorsRequest(attributes, clusterID, expiry), ActorRef.noSender());
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Override
+    public CompletableFuture<Map<TaskExecutorID, WorkerId>> getTaskExecutorWorkerMapping() {
+        return null;
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterAkkaImpl.java:185]: todo?

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/persistence/MantisJobStore.java [hunk 5]
```diff
@@ -169,6 +171,18 @@ public void storeNewTaskExecutor(TaskExecutorRegistration registration) throws I
         storageProvider.storeNewTaskExecutor(registration);
     }
 
+    public void storeNewDisabledTaskExecutorsRequest(DisableTaskExecutorsRequest request) throws IOException {
+
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/persistence/MantisJobStore.java:175]: TODO?

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java (hunk 1 above)
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java:82]: doc

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterAkkaImpl.java [hunk 6]
```diff
@@ -170,4 +172,21 @@ public CompletableFuture<TaskExecutorStatus> getTaskExecutorState(TaskExecutorID
                 .thenApply(TaskExecutorStatus.class::cast)
                 .toCompletableFuture();
     }
+
+    @Override
+    public CompletableFuture<Void> disabledTaskExecutorsAre(Map<String, String> attributes, Instant expiry) {
+        resourceClusterManagerActor.tell(
+            new ResourceClusterActor.DisableTaskExecutorsRequest(attributes, clusterID, expiry), ActorRef.noSender());
+        return CompletableFuture.completedFuture(null);
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterAkkaImpl.java:180]: why does this need to be a fut? Doesn't seem to actually have any fut to be used.

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java [hunk 7]
```diff
@@ -77,6 +79,12 @@ public interface ResourceCluster extends ResourceClusterGateway {
...
+
+    CompletableFuture<Map<TaskExecutorID, WorkerId>> getTaskExecutorWorkerMapping();
+
+    CompletableFuture<Map<TaskExecutorID, WorkerId>> getTaskExecutorWorkerMapping(String attribute, String attributeValue);
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java:86]: nit: input can be a set of attributes?

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java (hunk 1 above)
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java:82]: One thing I think requires further thinking is error/failure in upgrade workflow and what happens to this expiration. Othere then that the API lgtm.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java [hunk 8]
```diff
@@ -63,6 +66,8 @@
 import lombok.Value;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.flink.runtime.rpc.RpcService;
+import org.apache.flink.shaded.guava30.com.google.common.collect.Comparators;
+import org.apache.flink.util.Preconditions;
```
Comment (sundargates) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java:70]: use Preconditions from the guava package.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java [hunk 9]
```diff
@@ -73,37 +78,61 @@
  */
 @ToString(of = {"clusterID"})
 @Slf4j
-class ResourceClusterActor extends AbstractActorWithTimers {
+public class ResourceClusterActor extends AbstractActorWithTimers {
```
Comment (sundargates) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java:81]: why is this public suddenly? change this to package private.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/proto/DisableTaskExecutorsRequest.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+ * limitations under the License.
+ */
+
+package io.mantisrx.master.resourcecluster.proto;
+
+import java.util.Map;
+import lombok.Value;
+
+@Value
+public class DisableTaskExecutorsRequest {
```
Comment (sundargates) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/proto/DisableTaskExecutorsRequest.java:23]: Test deserialization of this.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/resourceprovider/ResourceClusterStorageProvider.java [hunk 11]
```diff
@@ -26,6 +26,7 @@
  * Interface for persisting resource cluster related data.
  * See {@link ResourceClusterStorageProviderAdapter} to see the binding/constructor requirements.
  */
+// todo(sundaram): move this to the persistence package
```
Comment (sundargates) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/resourceprovider/ResourceClusterStorageProvider.java:29]: get rid of this.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/persistence/SimpleCachedFileStorageProvider.java [hunk 12]
```diff
@@ -53,6 +56,8 @@
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.flink.util.Preconditions;
```
Comment (sundargates) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/persistence/SimpleCachedFileStorageProvider.java:60]: Use guava Preconditions.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/persistence/SimpleCachedFileStorageProvider.java [hunk 13]
```diff
@@ -68,11 +73,14 @@
...
-    private final static String ARCHIVE_DIR = "/tmp/MantisArchive";
     private static final Logger logger = LoggerFactory.getLogger(SimpleCachedFileStorageProvider.class);
-    private static final String JOB_CLUSTERS_DIR = SPOOL_DIR + "/jobClusters";
+
+    private final File spoolDir;
+    private final File archiveDir;
+    private final File jobClustersDir;
+    private final File resourceClustersDir;
+
+//    private static final String JOB_CLUSTERS_DIR = SPOOL_DIR + "/jobClusters";
```
Comment (sundargates) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/persistence/SimpleCachedFileStorageProvider.java:83]: get rid of this line.

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/TaskExecutorRegistration.java (hunk 2 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/TaskExecutorRegistration.java]: It should never be null. We ensure that even if the taskExecutorAttributes are not found in the JSON, the JsonSerializer should deserialize this field as an empty map.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterAkkaImpl.java (hunk 6 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterAkkaImpl.java]: Fixed this. Thanks for pointing it out.

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java (hunk 7 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java]: fixed.

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java (hunk 1 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java]: I expect this expiration to be set to a high value, such as a day after the command has been issued.

Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java]: done.

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java [hunk 14]
```diff
@@ -82,6 +86,31 @@ public interface ResourceCluster extends ResourceClusterGateway {
...
     CompletableFuture<Ack> refreshClusterScalerRuleSet();
 
+    /**
+     * Disables task executors that match the passed set of attributes
+     *
+     * @param attributes attributes that need to be present in the task executor's set of attributes.
+     * @param expiry     instant at which the request can be marked as complete. this is important because we cannot be constantly checking if new task executors match the disabled criteria or not.
+     * @return a future that completes when the underlying operation is registered by the system
+     */
+    CompletableFuture<Void> disableTaskExecutorsFor(Map<String, String> attributes, Instant expiry);
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java:96]: nit: will CompletableFuture<Ack> be more consistent?

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/TaskExecutorRegistration.java [hunk 15]
```diff
@@ -57,6 +67,28 @@ public class TaskExecutorRegistration {
...
+        this.taskExecutorAddress = taskExecutorAddress;
+        this.hostname = hostname;
+        this.workerPorts = workerPorts;
+        this.machineDefinition = machineDefinition;
+        this.taskExecutorAttributes = (taskExecutorAttributes == null) ? ImmutableMap.of() : taskExecutorAttributes;
+    }
+
+    public boolean containsAttributes(Map<String, String> requiredAttributes) {
+        return taskExecutorAttributes.entrySet().containsAll(requiredAttributes.entrySet());
+    }
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/TaskExecutorRegistration.java:90]: taskExecutorAttributes might be null on older version agents?

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java (hunk 8 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java:72]: Done

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/api/akka/route/v1/ResourceClustersNonLeaderRedirectRoute.java [hunk 16]
```diff
@@ -148,6 +152,11 @@ protected Route constructRoutes() {
                         post(() -> scaleClusterSku(clusterName))
                     ))
                 ),
+                path(
+                    PathMatchers.segment().slash("disableTaskExecutors"),
+                    (clusterName) -> pathEndOrSingleSlash(() -> concat(
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/api/akka/route/v1/ResourceClustersNonLeaderRedirectRoute.java:157]: nit: clusterID?

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java [hunk 17]
```diff
@@ -79,33 +84,57 @@ class ResourceClusterActor extends AbstractActorWithTimers {
...
+    public void preStart() throws Exception {
+        super.preStart();
+        List<DisableTaskExecutorsRequest> activeRequests =
+            mantisJobStore.loadAllDisableTaskExecutorsRequests(clusterID);
+        for (DisableTaskExecutorsRequest request : activeRequests) {
+            onNewDisableTaskExecutorsRequest(request);
+        }
+
+        timers().startPeriodicTimer(
+            "periodic-disabled-task-executors-test",
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java:135]: nit: periodic-disabled-task-executors-test to periodic-disabled-task-executors-{clusterID}

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java [hunk 18]
```diff
@@ -715,80 +969,50 @@ boolean onTaskExecutorStatusChange(TaskExecutorStatusChange statusChange) {
...
                 }
+                return true;
+            } else {
+                return false;
             }
-            return false;
         }
 
         @Nullable
         private WorkerId getWorkerId() {
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java:993]: nit: optional?

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java [hunk 19]
```diff
@@ -82,6 +86,31 @@ public interface ResourceCluster extends ResourceClusterGateway {
...
+    CompletableFuture<Map<TaskExecutorID, WorkerId>> getTaskExecutorWorkerMapping();
+
+    /**
+     * Gets the task executors to worker mapping for all task executors in the resource cluster that match the
+     * filtering criteria as represented by the attributes.
+     *
+     * @param attributes filtering criteria
+     * @return a future mapping task executor IDs to the work they are doing
+     */
+    CompletableFuture<Map<TaskExecutorID, WorkerId>> getTaskExecutorWorkerMapping(Map<String, String> attributes);
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java:112]: Is there API to know how many/which executors are being disabled atm? Also does getResourceOverview has data reflected on disabled nodes.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java [hunk 20]
```diff
@@ -116,6 +145,7 @@ public Receive createReceive() {
                 .match(GetRegisteredTaskExecutorsRequest.class, req -> sender().tell(getTaskExecutors(isRegistered), self()))
                 .match(GetBusyTaskExecutorsRequest.class, req -> sender().tell(getTaskExecutors(isBusy), self()))
                 .match(GetAvailableTaskExecutorsRequest.class, req -> sender().tell(getTaskExecutors(isAvailable), self()))
+                .match(GetDisabledTaskExecutorsRequest.class, req -> sender().tell(getTaskExecutors(isDisabled), self()))
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java:148]: One thing critical to this process is to disable or adjust the usage result to cluster scaler so it don't interfere with the uggrade process.

**Review by Andyz26 — APPROVED** [review]
T

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java (hunk 14 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java:96]: +1

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/TaskExecutorRegistration.java (hunk 15 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/TaskExecutorRegistration.java:90]: no, when it gets deserialized, the null automatically gets converted into an empty map. I have tests covering that.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java (hunk 17 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java:135]: nice find!

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java (hunk 18 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java:993]: not required I think given it's explicitly marked as nullable and is particularly used within the class.

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java (hunk 19 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/master/resourcecluster/ResourceCluster.java:112]: There's an API to get the information on how many executors are being disabled. I don't have an API to figure out which executors are disabled. I can add it though if you know the exact use-case.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java (hunk 20 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java:148]: +1

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java [hunk 21]
```diff
@@ -557,6 +694,142 @@ static class GetClusterUsageRequest {
...
+    private static class CheckDisabledTaskExecutors {
+        String reason;
+    }
+
+    @Value
+    static class GetTaskExecutorWorkerMappingRequest {
+        Map<String, String> attributes;
+    }
+
+    interface AvailabilityState {
```
Comment (sundargates) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java:707]: Add documentation.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java (hunk 9 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java]: fixed.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/proto/DisableTaskExecutorsRequest.java (hunk 10 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/proto/DisableTaskExecutorsRequest.java:23]: done.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/resourceprovider/ResourceClusterStorageProvider.java (hunk 11 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/resourceprovider/ResourceClusterStorageProvider.java]: done.

**Pull request merged**